compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -48,6 +52,42 @@ use rustc_hir::intravisit::Visitor;
 use std::cmp::{self, Ordering};
 use std::iter;
 
+#[derive(Debug)]
+struct ConfusablesDirective {
+    names: Vec<String>,
+}
+
+#[derive(Diagnostic)]
+#[diag(hir_typeck_invalid_rustc_confusable_attr)]
+struct ConfusablesDirectiveParseErr {
+    #[primary_span]
+    #[label]
+    span: Span,
+}
+
+impl ConfusablesDirective {
+    fn try_parse(attr: &NormalAttr) -> Result<Self, ConfusablesDirectiveParseErr> {
+        let AttrArgs::Delimited(DelimArgs { tokens, .. }) = &attr.item.args else {
+            return Err(ConfusablesDirectiveParseErr {
+                span: attr.item.span(),
+            });
+        };
+
+        let mut names = Vec::new();
+        for tok in tokens.trees() {
+            if let TokenTree::Token(tok, _) = tok
+                && let token::Token { kind, .. } = tok
+                && let token::TokenKind::Literal(lit) = kind
+                && let token::Lit { kind: token::LitKind::Str, symbol, .. } = lit
+            {
+                names.push(symbol.to_string());
+            }
+        }
+
+        Ok(ConfusablesDirective { names })
+    }
```

Comment:
See `rustc_attribute::builtin` on examples how to parse attributes. For instance `parse_stability`. The token tree should not be directly manipulated.